care/facility/api/serializers/patient_consultation.py

```diff
@@ -864,3 +869,80 @@ def validate(self, attrs):
     class Meta:
         model = PatientConsultation
         fields = ("email",)
+
+
+class PatientConsentSerializer(serializers.ModelSerializer):
+    id = serializers.CharField(source="external_id", read_only=True)
+    created_by = UserBaseMinimumSerializer(read_only=True)
+    archived_by = UserBaseMinimumSerializer(read_only=True)
+
+    class Meta:
+        model = PatientConsent
+
+        fields = (
+            "id",
+            "type",
+            "patient_code_status",
+            "archived",
+            "archived_by",
+            "archived_date",
+            "created_by",
+            "created_date",
+        )
+
+        read_only_fields = (
+            "id",
+            "created_by",
+            "created_date",
+            "archived",
+            "archived_by",
+            "archived_date",
+        )
+
+    def validate(self, attrs):
+        if attrs.get("type") == ConsentType.PATIENT_CODE_STATUS and not attrs.get(
+            "patient_code_status"
+        ):
+            raise ValidationError(
+                {
+                    "patient_code_status": [
+                        "This field is required for Patient Code Status Consent"
+                    ]
+                }
+            )
+
+        if attrs.get("type") != ConsentType.PATIENT_CODE_STATUS and attrs.get(
+            "patient_code_status"
+        ):
+            raise ValidationError(
+                {
+                    "patient_code_status": [
+                        "This field is not required for this type of Consent"
+                    ]
+                }
+            )
+        return attrs
+
+    def clear_existing_records(self, consultation, type, self_id=None):
+        consents = PatientConsent.objects.filter(
+            consultation=consultation, type=type
+        ).exclude(id=self_id)
+
+        # looping because .update does not call model save method
+        for consent in consents:
+            consent.archived = True
+            consent.archived_by = self.context["request"].user
+            consent.archived_date = timezone.now()
+            consent.save()
```

Comment:
we can use bulk update here

Reply:
bulk update does not trigger the model's save method, which is why had to resort to looping and saving 😅

Reply:
Oh, in that case you can do the operations in the save method, because this will cause (N consent archives) -> (NxN file archieves)